### PR TITLE
Bug Fix: API for fetching courses using school and code no longer working

### PR DESCRIPTION
**Pull request by DaleCS**
This is a bug fix. 

### The Bug:
Because the schema of course has been changed to break the course code into department and code. The existing API for fetching a course using their school and code no longer works.

### The Fix:
I simplified the 3 GET APIs that use the 'getPopulatedCourse' function into 1 GET API that uses optional queries to fetch courses.

This API is needed for fetching data of courses being added to a degree plan.

### CODE REVIEW

**Comment by calvinqc** [comment]
@DaleCS I'm thinking of this: Since we have 3 API that does the same job (getPopulatedCourse with different Options). So  I think we should convert these params into query, and do only 1 API /GET

File: client/src/app/course.service.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ export class CourseService {
       };
 
       this.allCourses = this.http.get<CourseData[]>(
-        `${this.uri}/course/` + school.toLowerCase(),
+        `${this.uri}/course?school=` + school.toUpperCase(),
```
Comment (DaleCS) [inline client/src/app/course.service.ts:42]: At this point I'm assuming that school will always be saved in our DB in uppercase.

File: server/controller/validation/course.validation.js [hunk 2]
```diff
@@ -57,3 +58,34 @@ export const validateCourseId = [
     .isLength({ min: 24, max: 24 })
     .withMessage(ID_IS_INVALID),
 ];
+
+export const validateCourseQuery = [
```
Comment (DaleCS) [inline server/controller/validation/course.validation.js:62]: Validation for the new 'validateCourseQuery' validation array that the new API uses.

File: server/controller/course.controller.js [hunk 3]
```diff
@@ -68,45 +69,19 @@ courseController.post(
...
+  validateCourseQuery,
   (req, res) => {
-    const school = req.params.school.toUpperCase();
     validationHandler(req, res, () => {
-      getPopulatedCourse({ school }, res);
+      const { _id, school, department, code } = req.query;
+      getPopulatedCourse(
+        {
+          ...(_id && { _id }),
+          ...(school && { school: school.toUpperCase() }),
```
Comment (DaleCS) [inline server/controller/course.controller.js:79]: Assuming all school field data and department are all uppercase.

File: client/src/app/course.service.ts (hunk 1 above)
Comment (calvinqc) [reply client/src/app/course.service.ts:42]: Yes, school will be SJSU for now

File: server/controller/course.controller.js [hunk 4]
```diff
@@ -68,45 +69,19 @@ courseController.post(
...
-  validateSchoolCourses,
+  validateCourseQuery,
   (req, res) => {
-    const school = req.params.school.toUpperCase();
     validationHandler(req, res, () => {
-      getPopulatedCourse({ school }, res);
+      const { _id, school, department, code } = req.query;
+      getPopulatedCourse(
+        {
+          ...(_id && { _id }),
```
Comment (calvinqc) [inline server/controller/course.controller.js:78]: This looks very nice, where did you learn this?

**Review by calvinqc — APPROVED** [review]
LGTM

File: server/controller/course.controller.js (hunk 4 above)
Comment (DaleCS) [reply server/controller/course.controller.js:78]: The second to the top answer of this stackoverflow post:
https://stackoverflow.com/questions/11704267/in-javascript-how-to-conditionally-add-a-member-to-an-object

**Pull request merged**